worker/provisioner/provisioner_task.go

```diff
@@ -1038,34 +1057,34 @@ func (task *provisionerTask) startMachine(
 		// If the failure of StartInstance() is due the availability zone,
 		// find a new one to use when retrying.
 		if errors.Cause(err) == environs.ErrAvailabilityZoneFailed {
-			if !distributeAcrossZones {
-				// User specified availability zone failed, nothing to retry.
-				task.removeMachineFromAZMap(machine)
-				return task.setErrorStatus("cannot start instance for machine %q in placement zone: %v", machine, err)
-			}
+			// Clean up AZMap from failed availability zone attempt
 			azRemaining, err2 := task.markMachineFailedInAZ(machine, startInstanceParams.AvailabilityZone)
 			if err2 != nil {
 				if err = task.setErrorStatus("cannot start instance: %v", machine, err2); err != nil {
 					logger.Errorf("setting error status: %s", err)
 				}
 				return err2
 			}
-			if azRemaining {
+
+			// Setup for next interation.
+			switch {
+			case azRemaining:
 				// There are more availability zones to try, don't burn an attempt.
-				retryMsg = fmt.Sprintf("failed to start instance (%s) within attempt %d, retrying in %v with new availability zone",
-					err.Error(),
-					task.retryStartInstanceStrategy.retryCount-attemptsLeft,
-					task.retryStartInstanceStrategy.retryDelay)
 				attemptsLeft++
-			} else {
+			case !azRemaining:
 				// All availability zones have been attempted for this iteration,
 				// clear the failures for the next time around.
 				task.clearMachineAZFailures(machine)
 			}
+
+			retryMsg = fmt.Sprintf("failed to start machine %s within attempt %d, retrying in %v with new availability zone",
```

Comment:
Is this right? We are setting the retryMsg even when there are no more zones to try?
Should we exit the loop when azRemaining is false? If we have tried all the zones on offer, why continue looping?

Reply:
If azRemaining is false, we clear the machine failures and so we can retry each zone.  Just because a zone has failed once, doesn't mean it won't succeeded a second time.  This also mimics the previous behavior.  Added an additional comment.